### PR TITLE
Implements the `start cvd server` stage.

### PR DESCRIPTION
**Pull request by ser-io**

### CODE REVIEW

**Review by jemoreira — CHANGES_REQUESTED** [review]

File: host/frontend/host-orchestrator/orchestrator/instancemanager.go [hunk 1]
```diff
@@ -107,6 +86,88 @@ func validateRequest(r *apiv1.CreateCVDRequest) error {
...
+func (p Procedure) Execute() error {
+	for _, s := range p {
+		if err := s.Run(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type LaunchCVDProcedureBuilding interface {
```
[the remark this answers is not in the document]
Comment (ser-io) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go]: done, used ProcedureBuilder.

File: host/frontend/host-orchestrator/orchestrator/instancemanager.go [hunk 2]
```diff
@@ -107,6 +86,88 @@ func validateRequest(r *apiv1.CreateCVDRequest) error {
 	return nil
 }
 
+func (m *InstanceManager) completeOperation(opName string, result OperationResult) {
```
[the remark this answers is not in the document]
Comment (ser-io) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go]: done

File: host/frontend/host-orchestrator/orchestrator/instancemanager.go [hunk 3]
```diff
@@ -107,6 +86,88 @@ func validateRequest(r *apiv1.CreateCVDRequest) error {
...
+type Procedure []ProcedureStage
+
+func (p Procedure) Execute() error {
+	for _, s := range p {
+		if err := s.Run(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (ser-io) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go:103]: created the `ProcedureBuilder` and `Procedure` abstractions to separate building from execution improving this way the testability of the code. now we can have tests checking the order of the stages and the stage instances on the building side, and we can have test for how these are executed.

I also find these design more extensible to new stages as we don't have to change the execution part, just add the new stage to the building part.

[the remark this answers is not in the document]
Comment (jemoreira) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go:103]: could you be more specific of what you can test now that you couldn't test before? some examples would help

File: host/frontend/host-orchestrator/orchestrator/instancemanager.go [hunk 4]
```diff
@@ -107,6 +86,88 @@ func validateRequest(r *apiv1.CreateCVDRequest) error {
...
+	Paths               IMPaths
+	CVDDownloader       *CVDDownloader
+	StartCVDServerCmd   StartCVDServerCmd
+	downloadCVDMutex    sync.Mutex
+	startCVDServerMutex sync.Mutex
+	cvdServerStarted    bool
+}
+
+func (b *LaunchCVDProcedureBuilder) Build(req apiv1.CreateCVDRequest) Procedure {
+	return []ProcedureStage{
```
[the remark this answers is not in the document]
Comment (ser-io) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go]: done

File: host/frontend/host-orchestrator/orchestrator/instancemanager.go [hunk 5]
```diff
@@ -35,62 +36,40 @@ func (s EmptyFieldError) Error() string {
...
-	return &InstanceManager{
-		rootDir:       rootDir,
-		om:            om,
-		cvdDownloader: cvdDownloader,
-		paths: IMPaths{
-			CVDBin: rootDir + "/cvd",
-		},
-	}
+	OM                         OperationManager
+	LaunchCVDProcedureBuilding LaunchCVDProcedureBuilding
```
[the remark this answers is not in the document]
Comment (ser-io) [reply host/frontend/host-orchestrator/orchestrator/instancemanager.go]: decided to have these fields exported and remove the existing constructor to benefit from the Golang `Composite literals` feature when creating new instances of the IM struct and  in this specific case having these fields exposed does not affect the internals or public contracts of the class.

**Review by jemoreira — APPROVED** [review]

**Pull request merged**